### DISCUSSION TITLE
Monitoring Model During Simulation

### DISCUSSION THREAD

**Discussion by a2l5t8** [cnrl/CoNeX#64] [Q&A] [answered]
When I simulate my network, I would like to know how much time and space is used for each component of my network in order to help optimize it, Is there any way to save the amount of time it takes for each behavior or each object like NeuronGroups to simulate without manually calculating these values?

**Comment by realamirhe** [comment, marked as answer]
Here's a code snippet I use to identify and troubleshoot bottlenecks:



```py
from pyinstrument import Profiler

with Profiler(async_mode="disabled") as p:
  run_your_conex_model_here()
  
p.print()
```

**Reply by a2l5t8** [reply]
in `run_your_conex_model_here()` do need to implement all the network definition or is it okay to only include only the following?
`
net.initialize()
net.simulate_iterations(ITERATIONS)
`

**Reply by realamirhe** [reply]
Cause the `net.simulate_iterations(ITERATIONS)` is the most time-consuming part, via its iterative nature, it suffices to only run network simulation within Profile with-context

**Comment by saeedark** [comment]
Pymonntorch's Network class has a method for measuring the blocks' time. 

https://github.com/cnrl/PymoNNtorch/blob/0f8243dac802fb340a558f2c1d2618ab5cbf7fb9/pymonntorch/NetworkCore/Network.py#L290

(Turn `measure_behavior_execution_time` to `True` for a single iteration.)
Due to the asynchronous nature of GPU execution, the measured times are not accurate, and therefore using a [profiler](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html)  is advised.
